File: foundations-jdbc/src/java/dev/typr/foundations/RowTemplate.java
package dev.typr.foundations;

import java.util.Arrays;
import java.util.Iterator;

public sealed interface RowTemplate<Row, Out> extends Template<Row, Out>
    permits RowTemplate.Query, RowTemplate.Update, RowTemplate.GeneratedKeys {

  record Query<Row, Out>(
      Fragment fragment,
      RowCodecNamed<Row> codec,
      int[] includedIndices,
      ResultSetParser<Out> resultParser)
      implements RowTemplate<Row, Out>, TemplateRead<Row, Out> {
    @Override
    public OperationRead.Query<Out> on(Row row) {
      Object[] encoded = codec.encode().apply(row);
      Object[] params = new Object[includedIndices.length];
      for (int i = 0; i < includedIndices.length; i++) params[i] = encoded[includedIndices[i]];
      return new OperationRead.Query<>(
          fragment.fill(Arrays.asList(params).iterator()), resultParser);
    }

    @Override
    public String description(boolean verbose) {
      return "RowTemplate.Query: " + fragment.renderInterpolated();
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Update<Row>(Fragment fragment, RowCodecNamed<Row> codec, int[] includedIndices)
      implements RowTemplate<Row, Integer> {
    @Override
    public Operation<Integer> on(Row row) {
      Object[] encoded = codec.encode().apply(row);
      Object[] params = new Object[includedIndices.length];
      for (int i = 0; i < includedIndices.length; i++) params[i] = encoded[includedIndices[i]];
      return new Operation.Update(fragment.fill(Arrays.asList(params).iterator()));
    }

    public Operation.UpdateManyTemplate<Row> onMany(Iterator<Row> rows) {
      return new Operation.UpdateManyTemplate<>(fragment, codec, includedIndices, rows);
    }

    @Override
    public String description(boolean verbose) {
      return "RowTemplate.Update: " + fragment.renderInterpolated();
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record GeneratedKeys<Row, Out>(
      Fragment fragment,
      RowCodecNamed<Row> codec,
      int[] includedIndices,
      String[] generatedColumns,
      ResultSetParser<Out> resultParser)
      implements RowTemplate<Row, Out> {
    @Override
    public Operation<Out> on(Row row) {
      Object[] encoded = codec.encode().apply(row);
      Object[] params = new Object[includedIndices.length];
      for (int i = 0; i < includedIndices.length; i++) params[i] = encoded[includedIndices[i]];
      return new Operation.UpdateReturningGeneratedKeys<>(
          fragment.fill(Arrays.asList(params).iterator()), generatedColumns, resultParser);
    }

    @Override
    public String description(boolean verbose) {
      return "RowTemplate.GeneratedKeys["
          + String.join(",", generatedColumns)
          + "]: "
          + fragment.renderInterpolated();
    }

    @Override
    public String toString() {
      return description(false);
    }
  }
}

File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/RowTemplate.kt
@file:Suppress("unused")
package dev.typr.foundationskt

sealed interface RowTemplate<Row : Any, Out> : Template<Row, Out> {

    class Query<Row : Any, Out>(
        override val underlying: dev.typr.foundations.RowTemplate.Query<Row, Out>
    ) : RowTemplate<Row, Out>, TemplateRead<Row, Out> {
        override fun on(input: Row): OperationRead.Query<Out> = OperationRead.Query(underlying.on(input))
    }

    class Update<Row : Any>(
        override val underlying: dev.typr.foundations.RowTemplate.Update<Row>
    ) : RowTemplate<Row, Int> {
        override fun on(input: Row): Operation<Int> = Operation.JavaWrapped(underlying.on(input))

        fun onMany(rows: Iterator<Row>): Operation.UpdateManyTemplate<Row> =
            Operation.UpdateManyTemplate(underlying.onMany(rows))
    }

    class GeneratedKeys<Row : Any, Out>(
        override val underlying: dev.typr.foundations.RowTemplate.GeneratedKeys<Row, Out>
    ) : RowTemplate<Row, Out> {
        override fun on(input: Row): Operation<Out> = Operation.JavaWrapped(underlying.on(input))
    }
}
